CMakeLists.txt

```diff
@@ -828,13 +806,15 @@ if(WITH_OPTIM)
         if(WITH_SSE42)
             check_sse42_intrinsics()
             if(HAVE_SSE42CRC_INLINE_ASM OR HAVE_SSE42CRC_INTRIN)
-                add_definitions(-DX86_SSE42)
                 set(SSE42_SRCS ${ARCHDIR}/adler32_sse42.c ${ARCHDIR}/insert_string_sse42.c)
                 add_feature_info(SSE42_CRC 1 "Support SSE4.2 optimized CRC hash generation, using \"${SSE42FLAG}\"")
                 list(APPEND ZLIB_ARCH_SRCS ${SSE42_SRCS})
                 set_property(SOURCE ${SSE42_SRCS} PROPERTY COMPILE_FLAGS "${SSE42FLAG} ${NOLTOFLAG}")
                 if(HAVE_SSE42CRC_INTRIN)
-                    add_definitions(-DX86_SSE42_CRC_INTRIN)
+                    add_definitions(-DX86_SSE42)
+                endif()
+                if(HAVE_SSE42CRC_INLINE_ASM)
+                    add_definitions(-DX86_SSE42_CRC_ASM)
```

Comment:
Zlib-ng required gcc 8+, clang 9+ and MSVC 2019+ (https://github.com/zlib-ng/zlib-ng/wiki).
All these compilers support the ``_mm_crc32_u32`` and ``__builtin_ia32_crc32si`` intrinsics.

We can simplify these checks by removing the unused assembler implementation.

Reply:
GCC has supported ``_mm_crc32_u32`` since at least version 4.4.
An implementation using ``__builtin_ia32_crc32si`` is redundant.

Reply:
> GCC has supported `_mm_crc32_u32` since at least version 4.4. An implementation using `__builtin_ia32_crc32si` is redundant.

This probably ended up this way due to gcc support being added first using the builtin and then later the MSC variant that uses the more standardized function and nobody actually checked whether GCC/Clang/etc supported that.

Reply:
When we started, a lot of people still used gcc 4.x, so it made sense to support all minor versions... Nowadays people use at least gcc 4.8, with a lot of people using at least gcc 6.x series. There is known issue with using gcc 7.x or later with some Xeon processors.

Reply:
> Zlib-ng required gcc 8+, clang 9+ and MSVC 2019+ (https://github.com/zlib-ng/zlib-ng/wiki). All these compilers support the `_mm_crc32_u32` and `__builtin_ia32_crc32si` intrinsics.
> 
> We can simplify these checks by removing the unused assembler implementation.

The asm fallback might in some cases still be valuable while cross-compiling or for compiling in a container. It is very much a corner case, but I don't think it hurts much to keep it around.

Reply:
> The asm fallback might in some cases still be valuable while cross-compiling or for compiling in a container. It is very much a corner case, but I don't think it hurts much to keep it around.

I think the assembler implementation makes the code more confusing, because is not used in practice.

Reply:
After mulling it over some more, I realize you are correct. It does get used in the cases I mentioned, but that does not matter because those are not acutally usable with --native anyways.